ignitionrobotics/ign-gazebo
# Add ability to record video based on sim time

The current default behavior of video recording in gui camera is to record using wall clock time as timestamp. This PR adds the ability to specify from the gui config to record videos using sim time. This is so that we can still produce smooth videos that run in real time even when the playback RTF is low or fluctuates often. To get perfect video timing, you'll also require this fix in ignitionrobotics/ign-common#105

To test:

1. add a `<record_video><use_sim_time>` flag to the scene3d plugin in your  `~/.ignition/gazebo/gui.config` (and maybe `~/.ignition/gazebo/playback_gui.config` if you're doing playback - for me I just had to update `gui.config`), e.g.

    ```
    
    <plugin filename="GzScene3D" name="3D View">
      <ignition-gui>
        <title>3D View</title>
        <property type="bool" key="showTitleBar">false</property>
        <property type="string" key="state">docked</property>
      </ignition-gui>

      <engine>ogre2</engine>
      <scene>scene</scene>
      <ambient_light>0.4 0.4 0.4</ambient_light>
      <background_color>0.8 0.8 0.8</background_color>
      <camera_pose>6 0 6 0 0.5 3.14</camera_pose>
      <record_video>
        <use_sim_time>true</use_sim_time>
      </record_video>
    </plugin>
    ```

1. Launch ign-gazebo with a complex scene so that the RTF is low. Alternatively, artificially increase cpu usage using `stress`, e.g. `stress -c 12`

1. Open the Video Recorder plugin (from top right menu) and start recording. Keep an eye on the duration of sim time that you record.

1. Stop the video recorder and save to file

1. Playback video and verify video length is the same as the duration of sim time that was recorded

Ideally this configuration is specified using a proper video record msg but adding that may introduce breaking changes.

Signed-off-by: Ian Chen <ichen@osrfoundation.org>
